src/Rect.h

```diff
@@ -6,6 +6,8 @@
 
 struct Rect
 {
+	Rect() : x1(0), y1(0), x2(0), y2(0) {}
+
```

Comment:
FYI: I believe the explicit constructor can affect a number of object properties that may be a bit unexpected. Such as what contexts a struct can be used, how initialization syntax works, or what default members are automatically provided. I'm a little fuzzy on the details, hence the vague warning this may have greater consequences than expected.

With that said, it's probably fine. Makes sense to initialize those values to 0.

I suppose if you really wanted to get into it, initialization could waste cycles if you're just going to immediately read into the object/overwrite it.

Reply:
Someone random on the internet's opinion: http://www.cs.technion.ac.il/users/yechiel/c++-faq/init-lists.html

I found when testing the version number of MapData, I would get inconsistent test results before initializing the value in the constructor. Sometimes, the compiler would chose a random number that was correct and sometimes it would chose a number that was incorrect. By explicitly setting values to 0 (or something else), it eliminates the possibility of receiving intermittent errors caused by randomly initialized values. I'm not aware of any limits this puts on the use of the class/struct.

For a struct like Rect, someone might decide to create 10,000 of them for legitimate reasons, so there could be a noticeable performance hit maybe?

I'm not hard set on doing this if it makes sense not to or only in certain cases. It just seemed better than letting the compiler chose wild and random numbers that could cause intermittent issues.

Brett

Reply:
Interesting read, thank you. I would agree that initializer lists can improve performance in some cases.

----

The point I brought up was more about having an explicit constructor, rather than the constructor using an initializer list. I did a little reading. Some relevant links are:
[Should I define the default constructor?](https://stackoverflow.com/questions/22184060/should-i-define-the-default-constructor)
[What are aggregates and PODs and how/why are they special?](https://stackoverflow.com/questions/4178175/what-are-aggregates-and-pods-and-how-why-are-they-special)

If you define a constructor, an automatic default constructor (having no parameters) is no longer provided. In this case, you're defining an explicit default constructor.

If you define a constructor, the type is no longer an aggregate. This affects brace initialization. You can initialize a struct with braces, unless you explicitly define a constructor, then it's an error:
```cpp
Rect rect { 1, 2, 5, 10 };  // Error with explicit default constructor
```

The above also affects array initialization:
```cpp
Rect rectArr[2] = {{1, 2, 5, 10}, {3, 4, 5, 10}}; // Error with explicit default constructor
```

----

The "randomly initialized" values are actually uninitialized values. The struct was being used before the fields were assigned values. That generally falls into the realm of undefined behaviour. If the compiler is able to detect use of uninitialized values, it will likely flag it with a warning or error. Unfortunately, there are cases where the compiler is not always able to detect this. In particular, the compiler likely has trouble with pointer and reference parameters, which could be in or out parameters. It's perfectly normal to pass uninitialized memory as an out parameter, as you may be expecting the function to initialize it. You wouldn't want to pass uninitialized memory as an in parameter.

Although random values from uninitialized fields is never good, I'm a little concerned this may be a sign of some deeper issue with the code. In particular, it may be trying to use a value before it should, or it may be skipping necessary error checks. Using zeros might be safer in the sense of no undefined behaviour, and likely no crash, but it may be just as wrong. Worse, it's now harder to detect because it doesn't crash or behave oddly.

Reply:
Thanks for providing the reading. I think there are pros and cons both ways, but we can stick with not defining default constructors for the POD like structs. 

While I agree that you should not rely on initializing values to 0 to hide bugs, it makes it easier to troubleshoot. Intermittent bugs are harder to deal with then repeatable results. Also, when I see a zero I know it was either not initialized (except by the constructor) or actually set to 0. If it is a random number I just don't know.